File: ruby/rpc_server.rb
#!/usr/bin/env ruby
require 'bunny'

class FibonacciServer
  def initialize(ch)
    @ch = ch
  end

  def start(queue_name)
    @q = @ch.queue(queue_name)
    @x = @ch.default_exchange

    @q.subscribe(block: true) do |dilivery_info, properties, payload|
      n = payload.to_i
      p "calculating fib(#{n})..."
      st = Time.now
      r = self.class.fib(n)
      p "sending result: #{r}, spent: #{Time.now - st}"

      @x.publish(
        r.to_s,
        routing_key: properties.reply_to,
        correlation_id: properties.correlation_id
      )
    end
  end

  def self.fib(n)
    case n
    when 0 then 0
    when 1 then 1
    else
      fib(n - 1) + fib(n - 2)
    end
  end
end

# create connection
conn = Bunny.new
conn.start

# create channel
ch = conn.create_channel

begin
  server = FibonacciServer.new(ch)
  puts ' [x] Awaiting RPC requests'
  server.start('rpc_queue')
rescue Interrupt => _exception
  puts ' [*] closing connection...'
  ch.close
  conn.close
  exit 0
end

File: elixir/rpc_server.exs
defmodule FibServer do
  def fib(0), do: 0
  def fib(1), do: 1
  def fib(n) when n > 1, do: fib(n-1) + fib(n-2)

  def wait_for_messages(channel) do
    receive do
      {:basic_deliver, payload, meta} ->
        {n, _} = Integer.parse(payload)
        st = Time.utc_now
        IO.puts " [.] calculating fib(#{n})..."
        response = fib(n)
        IO.puts " [.] sending result: #{response}, spent: #{Time.diff(Time.utc_now, st, :microsecond)/1000000}"

        AMQP.Basic.publish(
          channel,
          "",
          meta.reply_to,
          "#{response}",
          correlation_id: meta.correlation_id
        )
        AMQP.Basic.ack(channel, meta.delivery_tag)

        wait_for_messages(channel)
    end
  end
end

# create connection
{:ok, connection} = AMQP.Connection.open
{:ok, channel} = AMQP.Channel.open(connection)

# create channel
AMQP.Queue.declare(channel, "rpc_queue")

# subscribe to messages from queue
AMQP.Basic.qos(channel, prefetch_count: 1)
AMQP.Basic.consume(channel, "rpc_queue")
IO.puts " [x] Awaiting RPC requests"

FibServer.wait_for_messages(channel)
